MudMenu: Autocomplete/Select adornment click no longer causes menu to close

versile2 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
Added barriers at the outermostlevel for mousedown and click on MudSelect and MudAutocomplete. 
IActivator was catching the click event anytime an Adornment is Interactive. I added an exception clause for MudInputAdornments not to trigger IActivator. 

Exception class is `mud-no-activator`

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Existing Unit Tests, Visually, New Unit Tests for new method

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

versile2 committed:
updated Test Viewer

versile2 committed:
Prevents clicks and mousedowns  through autocomplete or select

versile2 committed:
add exception for input adornment for iactivatable

danielchalmers commented on src/MudBlazor/Base/MudBaseButton.cs:
```diff
@@ -128,7 +127,8 @@ protected virtual async Task OnClickHandler(MouseEventArgs ev)
                 return;
             await OnClick.InvokeAsync(ev);
-            Activatable?.Activate(this, ev);
+            if (!Class?.Contains("mud-input-adornment-icon-button") ?? true)
```
Does this tweak work? Might be easier to parse

```suggestion
            if (Class?.Contains("mud-input-adornment-icon-button") != true)
```

It doesn't seem right to add special conditions in the base button itself, but I don't fully get the IActivatable stuff. Would like to know what @ScarletKuro thinks

versile2 replied:
I was able to move the activator logic to MudMenu without much fanfare.

ScarletKuro replied:
> It doesn't seem right to add special conditions in the base button itself

I don't like it either, even if it's in MudMenu now. But I guess there is no other good ways? How we will deal with it when we will continue with removing `IActivatable`?

versile2 committed:
Relocate Activator logic to MudMenu

versile2 committed:
added unit test

versile2 committed:
additional unit test

danielchalmers commented:
Do you have a reproduction link I can try out

versile2 commented:
Updated try arc and made you a snippet. The Autocomplete used to close the menu when you clicked on the Adornment.

https://try.arctechonline.tech/snippet/GkQpERlMIsBJkxCQ

Yes appears to be the same, I could add another mud-ignore-activator class but mud-input-adornment needs to be there to cover textfield adornments, mudselect, mudautocomplete, heck any adornments shouldn't trigger menu if they are buttons (if they aren't buttons they are icons and it won't stop the Activator)

versile2 commented:
Also pretty sure IActivator gets to go in V9 so that helps.

danielchalmers commented:
@henon What are your thoughts on compatibility lists? https://github.com/MudBlazor/MudBlazor/pull/11011#issuecomment-2734361923

henon commented:
> @henon What are your thoughts on compatibility lists? [#11011 (comment)](https://github.com/MudBlazor/MudBlazor/pull/11011#issuecomment-2734361923)

Sorry, I don't understand neither, the question and the problem.

danielchalmers commented:
> > @henon What are your thoughts on compatibility lists? [#11011 (comment)](https://github.com/MudBlazor/MudBlazor/pull/11011#issuecomment-2734361923)
> 
> Sorry, I don't understand neither, the question and the problem.

Essentially fixing a case of #8287 by adding a special check inside MudMenu that ignores adornment buttons (`mud-input-adornment-icon-button`). @henon

henon commented:
> > > @henon What are your thoughts on compatibility lists? [#11011 (comment)](https://github.com/MudBlazor/MudBlazor/pull/11011#issuecomment-2734361923)
> > 
> > 
> > Sorry, I don't understand neither, the question and the problem.
> 
> Essentially fixing a case of #8287 by adding a special check inside MudMenu that ignores adornment buttons (`mud-input-adornment-icon-button`). @henon

Oh, I get it now. What about adding an empty `mud-no-activator` class to adornment button and check for this? This would allow us (and users) to use that class wherever needed to avoid driggering the IActivator.

danielchalmers commented:
> Oh, I get it now. What about adding an empty `mud-no-activator` class to adornment button and check for this? This would allow us (and users) to use that class wherever needed to avoid driggering the IActivator.

That could work 👍

versile2 committed:
Refactor MudInputAdornment and Clear activation behavior in reference to MudMenu

Updated `MudIconButton` to use "mud-no-activator" class inside of MudInput, preventing menu activation. Modified tests in `MenuTests.cs` to reflect this change and added a new test for regular buttons. Adjusted styles in `MudInput.razor.cs` and `MudInputAdornment.razor` to include the new class. Updated activation logic in `MudMenu.razor.cs` to ensure consistent behavior across components.

versile2 commented:
Done, changed Activator lookup to MudBaseButton, added class mud-no-activator as exclusion, added class to mud input activator button and updated tests to test both mudiconbutton or mudbutton for mud-no-activator.
